File: SWEA/java/hamburger.java
package javaTIL;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;


public class hamburger {
	static int material[];
	static int calo[];
	static int limitCal;
	static int maxScore;
	
	static void go(int depth, int sumScore, int sumCal) {
		
		if (sumScore > maxScore && sumCal <= limitCal)
			maxScore = sumScore;
		
		if (sumCal > limitCal)
			return;
		
		if (depth == material.length)
			return;
		
		go(depth+1, sumScore + material[depth], sumCal + calo[depth]);
		go(depth+1, sumScore, sumCal);	
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int T = Integer.parseInt(br.readLine());
		for (int tc = 1; tc <= T; tc++) {
			String[] str = br.readLine().split(" ");
			int materialNum = Integer.parseInt(str[0]);
			limitCal = Integer.parseInt(str[1]);
			material = new int[materialNum];
			calo = new int[materialNum];
			maxScore = 0;

			for (int i = 0; i < materialNum; i++) {
				String[] input = br.readLine().split(" ");	
				material[i] = Integer.parseInt(input[0]);
				calo[i] = Integer.parseInt(input[1]);
			}

			go(0,0,0);	
			System.out.println("#"+tc+" "+maxScore);
		}

	}
	
}

File: SWEA/python/Daily/D43/hamburger.py
import sys
sys.stdin = open("input.txt","r")

def DFS(depth, sumScore, sumCal):
    global maxScore
    if depth == materialNum:
        if sumScore > maxScore and sumCal <= limitCal:
            maxScore = sumScore
        return
    if sumCal > limitCal:
        return
    if sumScore > maxScore and sumCal <= limitCal:
        maxScore = sumScore
    DFS(depth+1,sumScore + scoreData[depth], sumCal + calData[depth])
    DFS(depth+1,sumScore, sumCal)

T = int(input())
for t in range(1, T+1):
    materialNum, limitCal = map(int, input().split())
    scoreData = [0] * materialNum
    calData = [0] * materialNum

    for i in range(materialNum):
        scoreData[i],calData[i] = map(int,input().split())
    maxScore = 0
    DFS(0,0,0)

    print(f"#{t} {maxScore}")







